File: frontend/src/api/question.ts
import apiClient, { unwrap } from './client'
import { sessionStore } from './session'

export interface Question {
  id: number
  content: string
  question_type: string
  options: string[]
  difficulty: string
  year?: number
  month?: number
  frequency: number
  subject_id: number
  subject_code?: string
  subject_name?: string
  chapter_id?: number
  score: number
  source?: string
  source_url?: string
  is_online?: boolean
  is_favorited?: boolean
  favorite_tags?: string[]
  favorite_note?: string
  answer?: string
  explanation?: string
}

export interface SearchQuestionsParams {
  keyword?: string
  subject_id?: number
  subject_code?: string
  subject_query?: string
  years?: number[]
  question_types?: string[]
  difficulty?: string
  chapter_ids?: number[]
  high_frequency?: boolean
  online_search?: boolean
  page?: number
  page_size?: number
}

export interface QuestionSearchResult {
  total: number
  page: number
  page_size: number
  local_count?: number
  online_count?: number
  online_enabled?: boolean
  items: Question[]
}

const serializeParams = (params: SearchQuestionsParams) => ({
  ...params,
  years: params.years?.join(','),
  question_types: params.question_types?.join(','),
  chapter_ids: params.chapter_ids?.join(','),
})

export const questionApi = {
  searchQuestions: (params: SearchQuestionsParams, signal?: AbortSignal) =>
    unwrap<QuestionSearchResult>(apiClient.get('/questions/search', { params: serializeParams(params), signal })),
  getQuestionDetail: (questionId: number) => unwrap<Question>(apiClient.get(`/questions/${questionId}`)),
  addToFavorites: (questionId: number, tags?: string[]) =>
    unwrap<{ success: boolean }>(apiClient.post('/questions/favorites', { question_id: questionId, tags })),
  removeFromFavorites: (questionId: number, userId = sessionStore.getUserId()) =>
    unwrap<{ success: boolean }>(apiClient.delete(`/questions/favorites/${userId}/${questionId}`)),
  getFavorites: (userId = sessionStore.getUserId(), page = 1, pageSize = 20) =>
    unwrap<QuestionSearchResult>(
      apiClient.get(`/questions/favorites/${userId}`, {
        params: { page, page_size: pageSize },
      }),
    ),
  getHighFrequencyQuestions: (subjectId: number, limit = 50) =>
    unwrap<Question[]>(apiClient.get(`/questions/high-frequency/${subjectId}`, { params: { limit } })),
}

File: backend/api/question.py
# -*- coding: utf-8 -*-
"""Question bank API routes."""

from typing import List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session

from backend.api.dependencies import get_current_user, get_optional_current_user
from backend.database import get_db
from backend.elasticsearch_client import es_client
from backend.models.user import User
from backend.services.question_service import QuestionService

router = APIRouter(prefix="/questions", tags=["questions"])


class AddToFavoritesRequest(BaseModel):
    question_id: int
    tags: Optional[List[str]] = None


def _split_ints(value: Optional[str]) -> Optional[List[int]]:
    if not value:
        return None
    return [int(item) for item in value.split(",") if item.strip()]


def _split_strings(value: Optional[str]) -> Optional[List[str]]:
    if not value:
        return None
    return [item.strip() for item in value.split(",") if item.strip()]


@router.get("/search")
async def search_questions(
    keyword: Optional[str] = None,
    subject_id: Optional[int] = None,
    subject_code: Optional[str] = None,
    subject_query: Optional[str] = None,
    years: Optional[str] = None,
    question_types: Optional[str] = None,
    difficulty: Optional[str] = None,
    chapter_ids: Optional[str] = None,
    high_frequency: Optional[bool] = None,
    online_search: bool = Query(default=True),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(get_optional_current_user),
):
    service = QuestionService(db, es_client)
    result = service.search_questions(
        keyword=keyword,
        subject_id=subject_id,
        subject_code=subject_code,
        subject_query=subject_query,
        years=_split_ints(years),
        question_types=_split_strings(question_types),
        difficulty=difficulty,
        chapter_ids=_split_ints(chapter_ids),
        high_frequency=high_frequency,
        online_search=online_search,
        page=page,
        page_size=page_size,
        user_id=current_user.id if current_user else None,
    )
    return {"code": 0, "data": result}


@router.get("/high-frequency/{subject_id}")
async def get_high_frequency_questions(
    subject_id: int,
    limit: int = Query(default=50, ge=1, le=100),
    db: Session = Depends(get_db),
):
    service = QuestionService(db, es_client)
    return {"code": 0, "data": service.get_high_frequency_questions(subject_id, limit)}


@router.post("/favorites")
async def add_to_favorites(
    request: AddToFavoritesRequest,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = QuestionService(db, es_client)
    success = service.add_to_favorites(current_user.id, request.question_id, request.tags)
    if not success:
        raise HTTPException(status_code=404, detail="Question not found.")
    return {"code": 0, "data": {"success": True}}


@router.get("/favorites/{user_id}")
async def get_favorites(
    user_id: int,
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    from backend.api.dependencies import require_self
    require_self(user_id, current_user)
    service = QuestionService(db, es_client)
    return {"code": 0, "data": service.get_favorites(user_id, page, page_size)}


@router.delete("/favorites/{user_id}/{question_id}")
async def remove_from_favorites(
    user_id: int,
    question_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    from backend.api.dependencies import require_self
    require_self(user_id, current_user)
    service = QuestionService(db, es_client)
    success = service.remove_from_favorites(user_id, question_id)
    if not success:
        raise HTTPException(status_code=404, detail="Favorite not found.")
    return {"code": 0, "data": {"success": True}}


@router.get("/{question_id}")
async def get_question_detail(
    question_id: int,
    db: Session = Depends(get_db),
    current_user: Optional[User] = Depends(get_optional_current_user),
):
    service = QuestionService(db, es_client)
    result = service.get_question_detail(question_id, user_id=current_user.id if current_user else None)
    if not result:
        raise HTTPException(status_code=404, detail="Question not found.")
    return {"code": 0, "data": result}
